cnd/src/btsieve/ethereum/mod.rs

```diff
@@ -24,11 +24,110 @@ use std::collections::HashSet;
-pub async fn matching_transaction<C>(
+pub async fn matching_deployed_transaction<'b, C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    _bytecode: Bytes,
+    _create_contract: CreateContract,
+) -> anyhow::Result<&'b Transaction>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    let (transaction, _) =
+        matching_transaction(blockchain_connector, start_of_swap, |_transaction| {
+            // TODO: Add logic for matching a deployed transaction.
+            (false, None)
+        })
+        .await?;
+
+    Ok(transaction)
+}
+
+pub async fn matching_funded_transaction<'b, C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    _location: Address,
+    _token_contract: Address,
+) -> anyhow::Result<(&'b Transaction, &'b TransactionReceipt)>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    match matching_transaction(blockchain_connector, start_of_swap, |_transaction| {
+        // TODO: Add logic for matching a funded transaction.
+        (false, None)
+    })
+    .await
+    {
+        Ok((transaction, Some(receipt))) => Ok((transaction, receipt)),
+        Ok((transaction, None)) => Err(anyhow::anyhow!(
+            "failed to get receipt for: {}",
+            transaction.hash
+        )),
+        Err(e) => Err(e),
+    }
+}
+
+pub async fn matching_refunded_transaction<'b, C>(
+    mut blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    _location: Address,
+) -> anyhow::Result<&'b Transaction>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    let (transaction, _) =
+        matching_transaction(blockchain_connector, start_of_swap, |_transaction| {
+            // TODO: Add logic for matching a refunded transaction.
+            (false, None)
+        })
+        .await?;
+
+    Ok(transaction)
+}
+
+pub async fn matching_redeemed_transaction<'b, C>(
+    mut blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    _location: Address,
+) -> anyhow::Result<(&'b Transaction, &'b TransactionReceipt)>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    match matching_transaction(blockchain_connector, start_of_swap, |_transaction| {
+        // TODO: Add logic for matching a redeemed transaction.
+        //
+        // Includes using the bloom filter to conditionally fetch receipts
```

Comment:
This is a very important!

How are you going to solve the issue of needing the receipt for refunded and redeemed?
In order for the underlying code to know that it has to fetch the receipt, you will kind of have to give it the event you are looking for.

The crucial code is here: https://github.com/comit-network/comit-rs/blob/64840879fae1d087d697a8c7530b527c757463a1/cnd/src/btsieve/ethereum/transaction_pattern.rs#L93-L95

Maybe it is actually enough to pass down the topic to matching_transactions? The underlying code can then filter out all blocks where the bloom filter doesn't match and call you for those that does match with a tuple of (`Transaction`, `Receipt`).

Reply:
Cheers, I won't be 100% sure until I flesh it out some more but I think the design can accommodate the bloom filter correctly - time will tell.